core/shared/src/main/scala/org/http4s/internal/CurlConverter.scala

```diff
@@ -64,18 +62,46 @@ private[http4s] object CurlConverter {
   def requestToCurlWithBody[F[_]: Concurrent](
       request: Request[F],
       redactHeadersWhen: CIString => Boolean,
-  )(implicit defaultCharset: Charset): F[(String, Request[F])] =
-    for {
-      cachedBodyStream <- request.body.compile[F, F, Byte].toVector.map(fs2.Stream.emits(_))
-      bodyString <- {
-        val cs = request.charset.getOrElse(defaultCharset).nioCharset
-        cachedBodyStream.through(text.decodeWithCharset(cs)).compile[F, F, String].string
+  )(implicit defaultCharset: Charset): F[(String, Request[F])] = {
+    val isBinary = request.contentType.exists(_.mediaType.binary)
+    val isJson = request.contentType.exists(mT =>
+      mT.mediaType == MediaType.application.json || mT.mediaType.subType.endsWith("+json")
+    )
+    val makeCurl: String => String =
+      bodyString => {
+        val bodyArg =
+          if (bodyString.isEmpty) "" else s"$newline--data ${PosixQuoting.quote(bodyString)}"
+        s"${request.asCurl(redactHeadersWhen)}$bodyArg"
       }
-    } yield {
-      val bodyArg =
-        if (bodyString.isEmpty) "" else s"$newline--data ${PosixQuoting.quote(bodyString)}"
-      val curl = s"${request.asCurl(redactHeadersWhen)}$bodyArg"
-      val refreshedRequest = request.withBodyStream(cachedBodyStream)
-      (curl, refreshedRequest)
+
+    request.entity match {
+      case Entity.Empty =>
+        val curl = makeCurl("")
+        Applicative[F].pure(curl -> request)
+
+      case Entity.Strict(b) =>
+        val bodyString =
+          if (!isBinary || isJson) {
+            val cs = request.charset.getOrElse(defaultCharset).nioCharset
+            b.decodeStringLenient()(cs)
+          } else
+            b.toHex
+        val curl = makeCurl(bodyString)
+        Applicative[F].pure(curl -> request)
+
+      case Entity.Streamed(body, _) =>
+        for {
+          cachedBodyStream <- body.compile[F, F, Byte].toVector.map(fs2.Stream.emits(_))
+          bodyString <- {
+            val cs = request.charset.getOrElse(defaultCharset).nioCharset
+            if (!isBinary || isJson)
+              cachedBodyStream.through(text.decodeWithCharset(cs)).compile[F, F, String].string
+            else
+              cachedBodyStream.compile[F, F, Byte].to(ByteVector).map(_.toHex)
```

Comment:
JFTR, aside from changes regarding the `Entity` model, I've added taking into account the Content-Type here. We similarly do in logging Middlewares.